src/main/java/net/querz/nbt/ListTag.java

```diff
@@ -411,93 +409,99 @@ public CompoundTag getCompoundOrDefault(int index, CompoundTag def) {
 	public ListTag getListOrDefault(int index, ListTag def) {
 		if (index >= 0 && index < value.size()) {
 			Tag tag = value.get(index);
-			if (tag.getID() == LIST) {
+			if (tag.getType() == LIST) {
 				return (ListTag) tag;
 			}
 		}
 		return def;
 	}
 
-	public <T extends Tag> List<T> iterateType(TagType<T> type) {
-		return new TypedListTag<>();
+	public <T extends Tag> List<T> iterateType(Class<T> tagClass) {
+		if (tagClass != type.tagClass) {
```

Comment:
this is also going to throw an NPE if `type` is null.

i think this should allow to iterate an empty list as well, at least it shouldn't throw an exception if we attempt to iterate an empty list. though it should throw an exception if we attempt to set or add a value with the wrong type.

so, this should be allowed:
```java
ListTag list = new ListTag();
for (ByteTag t : list.iterateType(ByteTag.class)) {
	System.out.println("hello there");
}
```
this should also be allowed:
```java
ListTag list = new ListTag();
List<ByteTag> typed = list.iterateType(ByteTag.class);
typed.add(ByteTag.valueOf((byte) 123));

System.out.println(NBTUtil.toSNBT(list)); // [123b]
```
this should not be allowed:
```java
ListTag list = new ListTag();
List<ByteTag> typed = list.iterateType(ByteTag.class);
list.add(IntTag.valueOf(123)); // still ok

for (ByteTag t : typed) { // throws ClassCastException
	System.out.println(NBTUtil.toSNBT(t));
}
```

Reply:
Whoops, you should definitely be able to iterate over an empty list yeah.

The other scenario's you listed are already covered by the current implementation of `TypedListTag`, but I've added some tests for good measure :)